packages/kit/src/runtime/server/page/actions.js

```diff
@@ -0,0 +1,197 @@
+import { error, json } from '../../../exports/index.js';
+import { negotiate } from '../../../utils/http.js';
+import { HttpError, Redirect, ValidationError } from '../../control.js';
+import { error_to_pojo } from '../utils.js';
+
+/**
+ * @param {import('types').RequestEvent} event
+ */
+export function is_action_json_request(event) {
+	console.log(event.request.headers.get('accept'));
+	const accept = negotiate(event.request.headers.get('accept') || 'text/html', [
+		'text/html',
+		'application/json'
+	]);
+
+	return (
+		accept === 'application/json' &&
+		event.request.method !== 'GET' &&
+		event.request.method !== 'HEAD'
+	);
+}
+
+/**
+ * @param {import('types').RequestEvent} event
+ * @param {import('types').SSROptions} options
+ * @param {import('types').SSRNode['server']} server
+ */
+export async function handle_action_json_request(event, options, server) {
+	// TODO create POJO interface for this? Something like
+	// {status: number, errors?: Record<string, any>, location?: string, values?: Record<string, any>, result?: Record<string, any>}
+	const actions = server.actions;
+
+	if (!actions) {
+		maybe_throw_migration_error(server);
+		// TODO should this be a different error altogether?
+		return new Response('POST method not allowed. No actions exist for this page', {
+			status: 405,
+			headers: {
+				// https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405
+				// "The server must generate an Allow header field in a 405 status code response"
+				allow: ''
+			}
+		});
+	}
+
+	try {
+		const result = await call_action(event, actions);
+		if (!result) {
+			// TODO json({status: 204}) instead?
+			return new Response(undefined, {
+				status: 204
+			});
+		} else {
+			return json({ status: 200, result });
+		}
+	} catch (e) {
+		const error = /** @type {Redirect | HttpError | ValidationError | Error} */ (e);
+
+		if (error instanceof Redirect) {
+			return json({ status: 303, location: error.location });
+		}
```

Comment:
We can't send 303s in the JSON case, because then the client has no idea what the URL is. I think we should send 200 responses with the relevant details encoded in the payload, and reserve non-200 status codes for non-validation errors

Reply:
This is a json 200 response with payload status/location

Reply:
🤦 this is what i get for reviewing code while tired